### ISSUE TITLE
Improvements Related to Namespace Package Portions for PathEntryFinders

### ISSUE THREAD

**Issue by ericsnowcurrently** [python/cpython#92896]
tl;dr there are some oddities we should consider addressing with namespace package portions coming from path entry finders.  It probably isn't worth bothering though (except for the docs fixes).

---

PEP 420 added (implicit) namespace packages.  At that time, and since then, namespace packages (and portions) have gotten special-casing throughout the import machinery.  Generally, we treat them like normal modules (or specs) even though certain characteristics don't apply.  This isn't a problem generally because namespace packages aren't that common (and because, I expect, only builtin importers ever deal with them).  However, there are a few things that it might be worth addressing.

PEP 451 introduced `ModuleSpec`, etc.  `PathEntryFinder.find_spec()` as the replacement for `PathEntryFinder.find_loader()`, with some notable differences:

* `find_loader()` returns `(loader, portion)`
   * the interpretation of this value is [clearly documented](https://docs.python.org/3/library/importlib.html#importlib.abc.PathEntryFinder.find_loader)
   * the finder can't provide any other possible information
   * if the returned loader is not `None` then the portion is ignored (not even a warning is set)
* `find_spec()` returns a `ModuleSpec` with `loader` set to `None` and `submodule_search_locations` set to the portion
   * the interpretation of this value for portions is not clearly documented anywhere
   * the path entry finder may set addition info on the spec (e.g. `loader_state`) but it is all ignored (not even a warning if set)

In both cases, [the language reference](https://docs.python.org/3/reference/import.html#namespace-packages) is not clear about how to identify namespace package portions from a path entry finder.

## What to Do?

* update the `PathEntryFinder.find_spec()` docs entry
* (a weak maybe) add the details to the language reference
* (maybe) be more explicit when returning a namespace package portion
* (maybe) strictly disallow (or warn) for extraneous information tied to a portion

### explicitly identify namespace package portions

1. do nothing
2. use a dedicated singleton instead of None
3. use a dedicated ModuleSpec subclass (e.g. NamespacePackagePortionSpec)
4. use a dedicated ModuleSpec-compatible class (e.g. NamespacePackagePortionSpec)

### explicitly deal with unexpectedly set data

(This is basically just if spec.origin is set.)

1. do nothing
2. emit a warning
3. fail

### explicitly deal with spec.loader_state

1. do nothing
2. emit a warning
3. fail
4. preserve the info on the spec of the combined namespace package

**Comment by ericsnowcurrently** [comment]
For reference, here's a summary of the code path for namespace packages:

* [`_find_and_load_unlocked()`](https://github.com/python/cpython/blob/37c6db60f9ac62b8a80bf04a8146274756ee0da0/Lib/importlib/_bootstrap.py#L1122)
    * [`PathFinder.find_spec()`](https://github.com/python/cpython/blob/37c6db6/Lib/importlib/_bootstrap_external.py#L1494)  # from sys.meta_path
        * [`PathFinder._get_spec()`](https://github.com/python/cpython/blob/37c6db6/Lib/importlib/_bootstrap_external.py#L1462)
            * for each path entry, call [`find_spec()`](https://github.com/python/cpython/blob/37c6db6/Lib/importlib/_bootstrap_external.py#L1598) on the corresponding path entry finder (FileFinder)
                * if [spec.loader is None](https://github.com/python/cpython/blob/37c6db6/Lib/importlib/_bootstrap_external.py#L1642) then spec.submodule_search_locations is a namespace package portion
                    * add it to any portions found earlier
                    * ignore any other data on the spec  # e.g. loader_state
                * return a new spec with loader set to None and submodule_search_locations set to the combined portions
        * return that spec after turning submodule_search_locations into a [_NamespacePath](https://github.com/python/cpython/blob/37c6db60f9ac62b8a80bf04a8146274756ee0da0/Lib/importlib/_bootstrap_external.py#L1264)
    * [`_load_unlocked()`](https://github.com/python/cpython/blob/37c6db60f9ac62b8a80bf04a8146274756ee0da0/Lib/importlib/_bootstrap.py#L666)
        * [`module_from_spec()`](https://github.com/python/cpython/blob/37c6db60f9ac62b8a80bf04a8146274756ee0da0/Lib/importlib/_bootstrap.py#L566)
            * [`_init_module_attrs()`](https://github.com/python/cpython/blob/37c6db60f9ac62b8a80bf04a8146274756ee0da0/Lib/importlib/_bootstrap.py#L507)
                * set spec.loader to a [NamespaceLoader](https://github.com/python/cpython/blob/37c6db60f9ac62b8a80bf04a8146274756ee0da0/Lib/importlib/_bootstrap_external.py#L1335)
        * NamepaceLoader.exec_module()  # a noop

(Note that #92279 will change the above slightly.)

**Comment by ericsnowcurrently** [comment]
CC @FFY00

I noticed this stuff while reviewing #92279.

**Comment by warsaw** [comment]
I vaguely remember diving into some related stuff here recently.

> find_spec() returns a ModuleSpec with loader set to None and submodule_search_locations set to the portion

Not quite.  Imports of *portions* -- i.e. submodules of a namespace package -- are nothing special; they are regular packages and have normal, expected specs with non-`None` loaders.  E.g. After creating a venv with Python 3.10 and `pip` installing both `flufl.i18n` and `flufl.lock`, where `flufl` is the namespace package and the former two are supplied by different portions.

```
>>> spec = find_spec('flufl.lock')
>>> spec
ModuleSpec(name='flufl.lock', loader=<_frozen_importlib_external.SourceFileLoader object at 0x1007fda20>, origin='/private/tmp/v2/lib/python3.10/site-packages/flufl/lock/__init__.py', submodule_search_locations=['/private/tmp/v2/lib/python3.10/site-packages/flufl/lock'])
```

Namespace packages themselves *are* special.  Did you mean that when you wrote "portion" above?  F.e. `__spec__.submodule_search_locations` on a namespace package will be an instance of a `_NamespacePath`, and that's also where you see `__spec__.loader = None`.

```
Python 3.10.4 (main, Apr 26 2022, 19:42:59) [Clang 13.1.6 (clang-1316.0.21.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from importlib.util import find_spec
>>> find_spec('flufl')
ModuleSpec(name='flufl', loader=None, submodule_search_locations=_NamespacePath(['/private/tmp/v2/lib/python3.10/site-packages/flufl']))
```

> the path entry finder may set addition info on the spec (e.g. loader_state) but it is all ignored (not even a warning if set)

Again, the details are murky, but I believe some finders do this to communicate additional information to the expected loader.  IIRC, this back channel is not well documented.

> In both cases, [the language reference](https://docs.python.org/3/reference/import.html#namespace-packages) is not clear about how to identify namespace package portions from a path entry finder.

I'm not sure why you need to identify namespace package *portions* as such, because they are just normal (sub)packages. Assuming you mean namespace packages, it's easy to identify them from the spec, even if it's undocumented and relies on a non-public symbol:

```
>>> from importlib._bootstrap_external import _NamespacePath
>>> spec = find_spec('flufl')
>>> isinstance(spec.submodule_search_locations, _NamespacePath)
True
```

I seem to remember a discussion about whether to publicly expose `_NamespacePath` (or create an alias for backward compatibility), remove the underscore, and document it, but we never did.

> update the PathEntryFinder.find_spec() docs entry

Yes.

> (a weak maybe) add the details to the language reference

Yes.

> (maybe) be more explicit when returning a namespace package portion

I don't think we need to do that.  Making `_NamespacePath` public is enough to unambiguously identify namespace packages.

> (maybe) strictly disallow (or warn) for extraneous information tied to a portion

Hmm.  Again, portions aren't special.

**Comment by warsaw** [comment]
Note that `loader=None` only happens if you call `find_spec()` before your import the namespace package.  I think that's implied by the [documentation](https://docs.python.org/3.11/library/importlib.html?highlight=find_spec#importlib.util.find_spec), but we could probably be clearer about that.

**Comment by warsaw** [comment]
```
Python 3.11.0b1+ (heads/3.11:65ebd923b3, May 18 2022, 09:30:39) [Clang 13.1.6 (clang-1316.0.21.2.3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from importlib.util import find_spec
>>> find_spec('flufl')
ModuleSpec(name='flufl', loader=None, submodule_search_locations=_NamespacePath(['/tmp/v3/lib/python3.11/site-packages/flufl']))
>>> import flufl
>>> find_spec('flufl')
ModuleSpec(name='flufl', loader=<_frozen_importlib_external.NamespaceLoader object at 0x1067e4b90>, submodule_search_locations=_NamespacePath(['/tmp/v3/lib/python3.11/site-packages/flufl']))
```

**Comment by ericsnowcurrently** [comment]
> Not quite. Imports of portions -- i.e. submodules of a namespace package -- are nothing special; they are regular packages and have normal, expected specs with non-None loaders.

Sorry for the confusion.  I'm not talking about importing the portions.  I'm talking about how PathFinder (the metapath finder) works.  For each path entry it calls the corresponding path entry finder's `find_spec()`, which returns either None or a module spec.  (The default path entry finder is FileFinder.)  The only special thing about a namespace package portion spec (coming from a path entry finder) is that `loader` is None.  PathFinder explicitly uses this fact to identify portions.  Once it does, it adds the portion spec's `submodule_search_paths` to a combined list, which later is wrapped in a _NamespacePath object.  Everything else from the portion's spec is thrown away.  (That's what caught my attention.)

I'm sure this is fine since FileFinder is probably the only path entry finder in existence that detects namespace package portions.  Furthermore, the only discarded data of significance is `origin` (should be None but we don't check) and `loader_state`.  The importance of `loader_state` depends entirely on the loader, and PathFinder doesn't care, so it doesn't really matter.  Aside from making this behavior clear in the docs, my only thought there is that is *might* be worth emitting a warning in PathFinder (in `_get_spec()`) if a portion spec has `origin` or `loader_state` set to anything other than None.

**Comment by warsaw** [comment]
> The only special thing about a namespace package portion spec (coming from a path entry finder) is that loader is None. PathFinder explicitly uses this fact to identify portions. Once it does, it adds the portion spec's submodule_search_paths to a combined list, which later is wrapped in a _NamespacePath object. Everything else from the portion's spec is thrown away. (That's what caught my attention.)

I think this explains why `find_spec('flufl')` pre-import returns different spec data than it does *after* `import flufl`.  Hmm, or maybe not, since `flufl` is a namespace package, not a portion.  It looks like portion spec's don't change after importing the submodule.

> Aside from making this behavior clear in the docs, my only thought there is that is might be worth emitting a warning in PathFinder (in _get_spec()) if a portion spec has origin or loader_state set to anything other than None.

+1 to documenting the behavior.  Probably +1 to the warning as well, but I don't have any sense of whether there exists anything that sets them to other than `None` or expects that to work.

**Comment by ericsnowcurrently** [comment]
> Probably +1 to the warning as well, but I don't have any sense of whether there exists anything that sets them to other than `None` or expects that to work.

Yeah, the number of folks doing anything custom with importers related to namespace packages is probably approximately zero. 😄  That said, the warning is an insignificant addition, would be very helpful if triggered, and would be as good as any comment we might put in that place in the code anyway.
